op-e2e/sgt/sgt_test.go

```diff
@@ -0,0 +1,307 @@
+		txValue         *big.Int
+		expectedErr     error
+		action          func(ctx context.Context, t *testing.T, index int64, sgtValue *big.Int, l2Value *big.Int, txValue *big.Int, sgt *SgtHelper) (*BalanceCheck, error)
+	}{
+		{
+			name:            "SGTDepositSuccess",
+			depositSgtValue: big.NewInt(10000000000000),
+			depositL2Value:  big.NewInt(0),
+			txValue:         big.NewInt(0),
+			action:          sgtDeposit,
+		},
+		{
+			name:            "NativaGasPaymentWithoutSGTSuccess",
+			depositSgtValue: big.NewInt(0),
+			depositL2Value:  big.NewInt(10000000000000),
+			txValue:         big.NewInt(0),
+			action:          sgtTxSuccess,
+		},
+		{
+			name:            "FullSGTGasPaymentWithoutNativeBalanceSuccess",
+			depositSgtValue: big.NewInt(10000000000000),
+			depositL2Value:  big.NewInt(0),
+			txValue:         big.NewInt(0),
+			action:          sgtTxSuccess,
+		},
+		{
+			name:            "FullSGTGasPaymentWithNativeBalanceSuccess",
+			depositSgtValue: big.NewInt(10000000000000),
+			depositL2Value:  big.NewInt(10000000000000),
+			txValue:         big.NewInt(0),
+			action:          sgtTxSuccess,
+		},
+		{
+			name:            "PartialSGTGasPaymentSuccess",
+			depositSgtValue: big.NewInt(1000),
+			depositL2Value:  big.NewInt(10000000000000),
+			txValue:         big.NewInt(0),
+			action:          sgtTxSuccess,
+		},
+		{
+			name:            "FullSGTGasPaymentAndNonZeroTxValueWithSufficientNativeBalanceSuccess",
+			depositSgtValue: big.NewInt(10000000000000),
+			depositL2Value:  big.NewInt(10000000000000),
+			txValue:         big.NewInt(10000),
+			action:          sgtTxSuccess,
+		},
+		{
+			name:            "PartialSGTGasPaymentAndNonZeroTxValueWithSufficientNativeBalanceSuccess",
+			depositSgtValue: big.NewInt(1000),
+			depositL2Value:  big.NewInt(10000000000000),
+			txValue:         big.NewInt(10000),
+			action:          sgtTxSuccess,
+		},
+		{
+			name:            "InsufficientGasPaymentFail",
+			depositSgtValue: big.NewInt(10000),
+			depositL2Value:  big.NewInt(10000),
+			txValue:         big.NewInt(0),
+			expectedErr:     errorInsufficientBalance,
+			action:          sgtTxFail,
+		},
+		{
+			name:            "FullSGTGasPaymentAndNonZeroTxValueWithInsufficientNativeBalanceFail",
+			depositSgtValue: big.NewInt(10000000000000),
+			depositL2Value:  big.NewInt(10000),
+			txValue:         big.NewInt(10000000000000),
+			expectedErr:     errorInsufficientBalance,
+			action:          sgtTxFail,
+		},
+		{
+			name:            "PartialSGTGasPaymentAndNonZeroTxValueWithInsufficientNativeBalanceFail",
+			depositSgtValue: big.NewInt(10000),
+			depositL2Value:  big.NewInt(10000),
+			txValue:         big.NewInt(10000000000000),
+			expectedErr:     errorInsufficientBalance,
+			action:          sgtTxFail,
+		},
+	}
+
+	for index, tCase := range tests {
+		t.Run(tCase.name, func(t *testing.T) {
+			b, err := tCase.action(ctx, t, int64(index), tCase.depositSgtValue, tCase.depositL2Value, tCase.txValue, sgt)
```

Comment:
Could you extract the common logic for handling `depositSgtValue`/`depositL2Value`/`txValue `?

Reply:
you mean, extract these into separate functions or methods?

Reply:
I was assuming the overall structure for each test is: 1. setup with `depositSgtValue/depositL2Value/txValue` 2. do its own stuff. So I think the setup logic should be reused. Is it the case?

Reply:
> I was assuming the overall structure for each test is: 1. setup with `depositSgtValue/depositL2Value/txValue` 2. do its own stuff. So I think the setup logic should be reused. Is it the case?

Yeah, the overall structure is correct. Most test cases are reusing the same code expect `sgtTxFail` and `sgtTxSuccess` functions. So, you mean the setup logic should be further reused in `sgtTxFail` and `sgtTxSuccess`?

Reply:
Yeah, and I think each entry action should be a unique action, instead of shared, like `sgtTxSuccess`, so that read knows what it's testing solely from the action name.

Reply:
done in https://github.com/ethstorage/optimism/pull/125/commits/2823c0fdf12e0329ed54e4bd3610d62822394a4b